st2common/st2common/util/param.py

```diff
@@ -170,17 +170,81 @@ def _process_defaults(G, schemas):
                 _process(G, name, value.get("default"))
 
 
+def _check_any_bad(G, nodes, check_any_bad=None, tracked_parents=None):
+    """
+    :param G: nx.DiGraph
+    :param nodes: list[dict]
+    :param check_any_bad: list[boolean]
+    :param tracked_parents: list[str]
+    """
+    if tracked_parents is None:
+        tracked_parents = []
+    if check_any_bad is None:
+        check_any_bad = []
+    for name in nodes:
+        if "value" not in G.nodes[name] and "template" not in G.nodes[name]:
+            # this is a string not a jinja template;  embedded {{sometext}}
+            check_any_bad.append(False)
+        else:
+            check_any_bad.append(True)
+        if name not in tracked_parents:
+            children = [i for i in G.predecessors(name)]
+            tracked_parents.extend(nodes)
+            _check_any_bad(G, children, check_any_bad, tracked_parents)
+    return check_any_bad
+
+
+def _remove_bad(g_copy, parent, nodes, tracked_parents=None):
+    """
+    :param G: nx.DiGraph
+    :param nodes: str
+    :param check_any_bad: list[str]
+    :param tracked_parents: list[str]
+    """
+
+    if tracked_parents is None:
+        tracked_parents = [parent]
+    else:
+        tracked_parents.append(parent)
+    for i in nodes:
+        if "template" in g_copy.nodes[parent].keys():
+            g_copy.nodes[parent]["value"] = g_copy.nodes[parent].pop("template")
+        if i in g_copy.nodes:
+            children = [i for i in g_copy.predecessors(i)]
+            if children:
+                if i not in tracked_parents:
```

Comment:
I don't see anywhere that adds the children to tracked_parents while walking the tree which means the `if i not in tracked_parents:` condition will always be true.